### PR TITLE
feat: aided_by tactic combinator for running tactics in the background

### PR DESCRIPTION
**Pull request by siddhartha-gadgil**
This implements a tactic combinator `aided_by` that automatically runs a tactic in the background. The syntax is essentially `"aided_by" tacticSeq "do" (tacticSeq)?`, with the first tactic sequence to be run in the background (e.g. `aesop?`) while the second tactic sequence is intractively entered. We also provide the sytax `"by!" tacticSeq` as a shorthand for `"aided_by" aesop? "do" tacticSeq`.

When elaborating each tactic in the (second) tactic sequence:
* We launch a tactic in a separate thread, which on completion caches the resulting proof state if successful.
* We wait briefly (50 ms by default) and then look for a proof state in the cache that matches the current goal and local context.
* If we find a matching proof state, we use it to complete the proof with a `TryThis` action.
* Otherwise the user continues proving interactively.
* Observe that as the user works interactively, the tactic sequence is re-evaluated, and hence we check if the background task completed the proof at some stage.
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by siddhartha-gadgil** [comment]
Demos on Zulip at https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Aesop.20etc.20in.20the.20background/near/378824560

**Comment by semorrison** [comment]
I'm a bit skeptical of an approach that relies on a fixed waiting time for the helper tactic to run. If it takes less, there's pointless delay, and if it takes more we just miss the answer? Neither is great.

I *very* much like the general idea of having tactics running automatically, but this doesn't feel like it is the right user interaction to drive it. I'm not sure I'll have time in the short term to think about this, so if others want to proceed here that's fine. But I'd prefer to think about this more.

**Comment by siddhartha-gadgil** [comment]
> I'm a bit skeptical of an approach that relies on a fixed waiting time for the helper tactic to run. If it takes less, there's pointless delay, and if it takes more we just miss the answer? Neither is great.
> 
> I _very_ much like the general idea of having tactics running automatically, but this doesn't feel like it is the right user interaction to drive it. I'm not sure I'll have time in the short term to think about this, so if others want to proceed here that's fine. But I'd prefer to think about this more.

@semorrison The idea is not to have a fixed waiting time, at least as the main interface. The tactic keeps running in the background even after that time and if the result is found it records this. Then when the same tactic sequence is elaborated again (for example when continuing to prove), the result is offered.

The waiting time is a very short one only for quick disposal.

File: Mathlib/Tactic/AidedBy.lean [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        logInfo m!"failed to parse tactic ({msg.toString})"
+        pure ()
+  return (tac, msgs)
+
+/--
+Run a tactic in `Meta` and cache the resulting proof state
+if the tactic is successful and finishing.
+-/
+def runAndCacheM (tacticCode : TSyntax ``tacticSeq)
+  (goal: MVarId) (target : Expr)  : MetaM Unit :=
```
[the remark this answers is not in the document]
Comment (BoltonBailey) [reply Mathlib/Tactic/AidedBy.lean]: I'm going to resolve these, they seem to be fixed.

**Comment by BoltonBailey** [comment]
Playing around with this I notice that changing `by` to `by#` in some places causes the tactic proof to fail where it was previously succeeding. The first example I found of this was `support_sum_eq_biUnion` in `Algebra.BigOperators.Finsupp`. If the `aided_by` mode is only working in the background, shouldn't complete proofs be unaffected by it?

**Comment by siddhartha-gadgil** [comment]
> Playing around with this I notice that changing `by` to `by#` in some places causes the tactic proof to fail where it was previously succeeding. The first example I found of this was `support_sum_eq_biUnion` in `Algebra.BigOperators.Finsupp`. If the `aided_by` mode is only working in the background, shouldn't complete proofs be unaffected by it?

May be bugs fixed in the newer implementation. Will take a look.

**Comment by siddhartha-gadgil** [comment]
A separate repository [LeanAideTools](https://github.com/siddhartha-gadgil/LeanAideTools) has been made to allow this to be used. If the mode proves useful a fresh PR with the new code will be opened.

**Pull request closed without merging**